lib/vsdata.cpp

```diff
@@ -47,6 +47,60 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int order = gf.FESpace()->GetMaxElementOrder();
+
+   // check for integral elements
+   const int dim = mesh->Dimension();
+   const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+   if (fec && fec->GetMapType(dim) == FiniteElement::INTEGRAL)
+   {
+      cout << "Warning: integral elements are non-polynomial in the physical space,\n"
+           << "         consider increasing the refinement by the key 'o'."
+           << endl;
```

Comment:
Showing this message here may be a bit excessive -- it will be shown regardless of the actual refinement factor too. It may be better to show it just when the solution grid function is set.

Treat this as a recommendation -- I'll also fine if you decide to keep it here, as it is.

Reply:
As I mentioned, even piece-wise constant integral elements on bilinear quads can be very curved (in the sense of the value-lifted surface), so the only safe combination is piece-wise constant integral elements on linear triangles. I do not know, it is a very special case, so I would not complicate it, but I may add it if you really want. The reason for placing it here is that it shows just next to the refinement level in console, so you see what it refers to 😉 .